GEECS-PythonAPI/geecs_python_api/controls/data_acquisition/scan_manager.py

```diff
@@ -1175,7 +1176,11 @@ def _execute_step(self, component_vars, wait_time, is_composite, max_retries=3,
 
 
             time.sleep(interval_time)
-            current_time += interval_time
+            current_time = time.time() - start_time
+
+            if current_time % 1 < interval_time:
+                log_df = self.scan_data_manager._process_results(self.results)
+                self.scan_data_manager.dataframe_to_tdms(log_df)
```

Comment:
I don't think you need the 'self.scan_data_manager.dataframe_to_tdms(log_df)' bit of code. That's already in the _process_results method. 

I'm not sure I understand this logic: if current_time % 1 < interval_time...

I get the idea here (just write the file while doing the steps), but maybe not the best implementation. I think effectively this is just dumping the file at various time steps. So, just overwriting existing files. What I think tony was after was to write the .tdms file in a similar fashion to MC. In MC, the file gets opened and written to continuously. And it is also possible to read/write to this file simultaneously. 

It all depends a little bit on what is trying to be achieved with the live dumping of the data files. Again, i don't quite follow the logic, but if you tested it and it works, this is a pretty brute force way to get the data files to update and maybe that's enough...

Reply:
We saw the `_process_results` method, but since they only wanted the tdms file to be dumped every shot I figured it would be best to save only the tdms and not also the txt file.  Otherwise, like you said this current implementation is pretty brute force and so writing more than we have to might not be efficient.

The `current_time % 1 < interval_time:` statement is just so that this code only executes once per second.  Otherwise it will get called every 0.1 seconds and that is too much.

I tested it with 1 camera, so we'll have to see what happens when there's a lot more going on in a scan.  Might be a significant chunk of time.  Would be good to debug just how long it takes to write a tdms file for a scan with 10 devices and 1000+ shots.  Might even want to consider a flag to turn off this feature if we plan on doing long, overnight scans.

Perhaps the biggest issue with leaving the tdms file open during the full run is the possibility of a crash occurring and then it gets corrupted?  But if that's not a problem then yeah we can revisit this at some point.

Reply:
my point on _process_results is that this is already in that method:
            # Write TDMS files (data and index)
            self.dataframe_to_tdms(log_df)
so no need to call it again after calling _process_results() as you have it on 1183.

I think I take back my point about overwriting the files. The tdms_writer stays open. So each time you call it, it is writing the full current data. I'm not sure there is anything that prevents you from writing duplicate entries in a tdms, so you might want to look at if you are just appending the full available dataframe at each step. 

I guess I understand the idea behind the code below now. But, I feel like it's not very deterministic. With current_time updating each time through the loop, in principle you would expect it to only increment by the interval_time. But, with other threads and processes going on, isn't it possible execution of that loop might take more than the interval_time? In which case, the decimal part of the number is going to be kind of random. 
            current_time = time.time() - start_time

            if current_time % 1 < interval_time: